File: python/bam_parser.py
"""
BAM Parser module for scMicrobe.

Extracts cell barcodes, UMIs, and candidate reads from scRNA-seq BAM files.
"""

import logging
from typing import Iterator, Dict, List, Optional, Tuple
from dataclasses import dataclass
import pysam
from tqdm import tqdm


@dataclass
class ReadRecord:
    """Represents a single read record with metadata."""
    read_id: str
    sequence: str
    quality: str
    barcode: Optional[str]
    umi: Optional[str]
    mapq: int
    is_unmapped: bool
    is_secondary: bool
    is_supplementary: bool
    reference_name: Optional[str] = None
    reference_start: Optional[int] = None
    cigar: Optional[str] = None
    
    def to_fasta(self) -> str:
        """Convert to FASTA format."""
        return f">{self.read_id}\n{self.sequence}\n"
    
    def to_fastq(self) -> str:
        """Convert to FASTQ format."""
        return f"@{self.read_id}\n{self.sequence}\n+\n{self.quality}\n"


class BAMParser:
    """Parser for scRNA-seq BAM files."""
    
    def __init__(
        self,
        bam_path: str,
        barcode_tag: str = 'CB',
        umi_tag: str = 'UB',
        chunk_size: int = 10000
    ):
        """
        Initialize BAM parser.
        
        Args:
            bam_path: Path to input BAM file
            barcode_tag: BAM tag for cell barcode (default: CB)
            umi_tag: BAM tag for UMI (default: UB)
            chunk_size: Number of reads per chunk for processing
        """
        self.bam_path = bam_path
        self.barcode_tag = barcode_tag
        self.umi_tag = umi_tag
        self.chunk_size = chunk_size
        self.logger = logging.getLogger('BAMParser')
        
        # Statistics
        self.stats = {
            'total_reads': 0,
            'mapped_reads': 0,
            'unmapped_reads': 0,
            'low_mapq_reads': 0,
            'reads_with_barcode': 0,
            'reads_with_umi': 0
        }
    
    def _extract_barcode(self, read) -> Optional[str]:
        """Extract cell barcode from read."""
        if self.barcode_tag and read.has_tag(self.barcode_tag):
            barcode = read.get_tag(self.barcode_tag)
            # Remove '-1' suffix if present
            if barcode and '-' in barcode:
                barcode = barcode.split('-')[0]
            return barcode
        return None
    
    def _extract_umi(self, read) -> Optional[str]:
        """Extract UMI from read."""
        if self.umi_tag and read.has_tag(self.umi_tag):
            return read.get_tag(self.umi_tag)
        return None
    
    def _read_to_record(self, read) -> ReadRecord:
        """Convert pysam read to ReadRecord."""
        return ReadRecord(
            read_id=read.query_name,
            sequence=read.query_sequence or '',
            quality=pysam.qualities_to_qualitystring(read.query_qualities) if read.query_qualities else '',
            barcode=self._extract_barcode(read),
            umi=self._extract_umi(read),
            mapq=read.mapping_quality,
            is_unmapped=read.is_unmapped,
            is_secondary=read.is_secondary,
            is_supplementary=read.is_supplementary,
            reference_name=read.reference_name if not read.is_unmapped else None,
            reference_start=read.reference_start if not read.is_unmapped else None,
            cigar=read.cigarstring if not read.is_unmapped else None
        )
    
    def extract_candidate_reads(
        self,
        mapq_threshold: int = 30,
        include_unmapped: bool = True,
        include_low_mapq: bool = True
    ) -> List[ReadRecord]:
        """
        Extract candidate microbial reads from BAM.
        
        Candidate reads are those that:
        1. Are unmapped, OR
        2. Have low MAPQ (< threshold), OR
        3. Are not confidently mapped to host
        
        Args:
            mapq_threshold: MAPQ threshold for confident host alignment
            include_unmapped: Include unmapped reads
            include_low_mapq: Include low MAPQ reads
            
        Returns:
            List of ReadRecord objects
        """
        candidates = []
        
        self.logger.info(f"Parsing BAM file: {self.bam_path}")
        
        with pysam.AlignmentFile(self.bam_path, 'rb') as bam:
            # Try to get total reads, fallback to None if index not available
            try:
                total_reads = bam.mapped + bam.unmapped
            except ValueError:
                total_reads = None
            
            for read in tqdm(bam, total=total_reads, desc="Parsing BAM"):
                self.stats['total_reads'] += 1
                
                # Skip secondary and supplementary alignments
                if read.is_secondary or read.is_supplementary:
                    continue
                
                # Track statistics
                if read.is_unmapped:
                    self.stats['unmapped_reads'] += 1
                else:
                    self.stats['mapped_reads'] += 1
                    if read.mapping_quality < mapq_threshold:
                        self.stats['low_mapq_reads'] += 1
                
                # Check if read has barcode
                barcode = self._extract_barcode(read)
                if barcode:
                    self.stats['reads_with_barcode'] += 1
                
                # Check if read has UMI
                if self._extract_umi(read):
                    self.stats['reads_with_umi'] += 1
                
                # Determine if this is a candidate read
                is_candidate = False
                
                if include_unmapped and read.is_unmapped:
                    is_candidate = True
                elif include_low_mapq and not read.is_unmapped and read.mapping_quality < mapq_threshold:
                    is_candidate = True
                
                if is_candidate:
                    record = self._read_to_record(read)
                    candidates.append(record)
        
        self._log_stats()
        return candidates
    
    def extract_reads_by_barcode(
        self,
        barcodes: List[str]
    ) -> Iterator[List[ReadRecord]]:
        """
        Extract reads for specific cell barcodes.
        
        Args:
            barcodes: List of cell barcodes to extract
            
        Yields:
            Chunks of ReadRecord objects
        """
        barcode_set = set(barcodes)
        chunk = []
        
        with pysam.AlignmentFile(self.bam_path, 'rb') as bam:
            for read in bam:
                barcode = self._extract_barcode(read)
                if barcode and barcode in barcode_set:
                    chunk.append(self._read_to_record(read))
                    
                    if len(chunk) >= self.chunk_size:
                        yield chunk
                        chunk = []
        
        if chunk:
            yield chunk
    
    def get_unique_barcodes(self) -> set:
        """Get all unique cell barcodes from BAM file."""
        barcodes = set()
        
        with pysam.AlignmentFile(self.bam_path, 'rb') as bam:
            for read in bam:
                barcode = self._extract_barcode(read)
                if barcode:
                    barcodes.add(barcode)
        
        return barcodes
    
    def _log_stats(self):
        """Log parsing statistics."""
        self.logger.info("BAM Parsing Statistics:")
        self.logger.info(f"  Total reads: {self.stats['total_reads']:,}")
        self.logger.info(f"  Mapped reads: {self.stats['mapped_reads']:,}")
        self.logger.info(f"  Unmapped reads: {self.stats['unmapped_reads']:,}")
        self.logger.info(f"  Low MAPQ reads: {self.stats['low_mapq_reads']:,}")
        self.logger.info(f"  Reads with barcode: {self.stats['reads_with_barcode']:,}")
        self.logger.info(f"  Reads with UMI: {self.stats['reads_with_umi']:,}")
    
    def write_to_fastq(
        self,
        reads: List[ReadRecord],
        output_path: str
    ):
        """Write reads to FASTQ file."""
        with open(output_path, 'w') as f:
            for read in reads:
                f.write(read.to_fastq())
        self.logger.info(f"Written {len(reads)} reads to {output_path}")


def main():
    """Test the BAM parser."""
    import sys
    
    if len(sys.argv) < 2:
        print("Usage: python bam_parser.py <bam_file>")
        sys.exit(1)
    
    logging.basicConfig(level=logging.INFO)
    
    parser = BAMParser(sys.argv[1])
    candidates = parser.extract_candidate_reads()
    
    print(f"Found {len(candidates)} candidate reads")
    
    if candidates:
        print("\nFirst 5 candidates:")
        for i, read in enumerate(candidates[:5]):
            print(f"  {i+1}. {read.read_id} | Barcode: {read.barcode} | UMI: {read.umi}")


if __name__ == '__main__':
    main()

File: go/bam_parser.go
package main

import (
	"bufio"
	"compress/gzip"
	"encoding/csv"
	"flag"
	"fmt"
	"log"
	"os"
	"runtime"
	"strings"
	"sync"

	"github.com/biogo/hts/bam"
	"github.com/biogo/hts/sam"
)

// ReadRecord represents a single read with metadata
type ReadRecord struct {
	ReadID          string
	Sequence        string
	Quality         string
	Barcode         string
	UMI             string
	MapQ            int
	IsUnmapped      bool
	IsSecondary     bool
	IsSupplementary bool
	ReferenceName   string
	ReferenceStart  int
}

// ToFastq converts record to FASTQ format
func (r *ReadRecord) ToFastq() string {
	return fmt.Sprintf("@%s\n%s\n+\n%s\n", r.ReadID, r.Sequence, r.Quality)
}

// ToFasta converts record to FASTA format
func (r *ReadRecord) ToFasta() string {
	return fmt.Sprintf(">%s\n%s\n", r.ReadID, r.Sequence)
}

// BAMParser parses scRNA-seq BAM files
type BAMParser struct {
	bamPath     string
	barcodeTag  string
	umiTag      string
	chunkSize   int
	numWorkers  int
}

// NewBAMParser creates a new BAM parser
func NewBAMParser(bamPath, barcodeTag, umiTag string, chunkSize, numWorkers int) *BAMParser {
	if numWorkers <= 0 {
		numWorkers = runtime.NumCPU()
	}
	return &BAMParser{
		bamPath:    bamPath,
		barcodeTag: barcodeTag,
		umiTag:     umiTag,
		chunkSize:  chunkSize,
		numWorkers: numWorkers,
	}
}

// extractBarcode extracts cell barcode from SAM record
func (p *BAMParser) extractBarcode(rec *sam.Record) string {
	for _, aux := range rec.AuxFields {
		tag := aux.Tag()
		if string(tag[:]) == p.barcodeTag {
			val := strings.TrimPrefix(aux.String(), p.barcodeTag+":Z:")
			// Remove -1 suffix if present
			if idx := strings.Index(val, "-"); idx != -1 {
				val = val[:idx]
			}
			return val
		}
	}
	return ""
}

// extractUMI extracts UMI from SAM record
func (p *BAMParser) extractUMI(rec *sam.Record) string {
	for _, aux := range rec.AuxFields {
		tag := aux.Tag()
		if string(tag[:]) == p.umiTag {
			return strings.TrimPrefix(aux.String(), p.umiTag+":Z:")
		}
	}
	return ""
}

// recordToStruct converts SAM record to ReadRecord
func (p *BAMParser) recordToStruct(rec *sam.Record) *ReadRecord {
	refName := ""
	refStart := 0
	if rec.Ref != nil {
		refName = rec.Ref.Name()
		refStart = rec.Start()
	}

	seq := ""
	if rec.Seq.Length > 0 {
		seqBytes := rec.Seq.Expand()
		seq = string(seqBytes)
	}

	qual := ""
	if len(rec.Qual) > 0 {
		qual = string(rec.Qual)
	}

	return &ReadRecord{
		ReadID:          rec.Name,
		Sequence:        seq,
		Quality:         qual,
		Barcode:         p.extractBarcode(rec),
		UMI:             p.extractUMI(rec),
		MapQ:            int(rec.MapQ),
		IsUnmapped:      rec.Flags&sam.Unmapped != 0,
		IsSecondary:     rec.Flags&sam.Secondary != 0,
		IsSupplementary: rec.Flags&sam.Supplementary != 0,
		ReferenceName:   refName,
		ReferenceStart:  refStart,
	}
}

// Stats holds parsing statistics
type Stats struct {
	TotalReads       int64
	MappedReads      int64
	UnmappedReads    int64
	LowMapQReads     int64
	ReadsWithBarcode int64
	ReadsWithUMI     int64
	CandidateReads   int64
}

// ExtractCandidateReads extracts candidate microbial reads
func (p *BAMParser) ExtractCandidateReads(mapqThreshold int, includeUnmapped, includeLowMapQ bool) ([]*ReadRecord, *Stats, error) {
	f, err := os.Open(p.bamPath)
	if err != nil {
		return nil, nil, fmt.Errorf("opening BAM: %w", err)
	}
	defer f.Close()

	br, err := bam.NewReader(f, p.numWorkers)
	if err != nil {
		return nil, nil, fmt.Errorf("creating BAM reader: %w", err)
	}
	defer br.Close()

	stats := &Stats{}
	var candidates []*ReadRecord

	for {
		rec, err := br.Read()
		if err != nil {
			break
		}

		stats.TotalReads++

		// Skip secondary and supplementary alignments
		if rec.Flags&sam.Secondary != 0 || rec.Flags&sam.Supplementary != 0 {
			continue
		}

		// Track statistics
		if rec.Flags&sam.Unmapped != 0 {
			stats.UnmappedReads++
		} else {
			stats.MappedReads++
			if int(rec.MapQ) < mapqThreshold {
				stats.LowMapQReads++
			}
		}

		// Check barcode and UMI
		if p.extractBarcode(rec) != "" {
			stats.ReadsWithBarcode++
		}
		if p.extractUMI(rec) != "" {
			stats.ReadsWithUMI++
		}

		// Determine if candidate
		isCandidate := false
		if includeUnmapped && rec.Flags&sam.Unmapped != 0 {
			isCandidate = true
		} else if includeLowMapQ && rec.Flags&sam.Unmapped == 0 && int(rec.MapQ) < mapqThreshold {
			isCandidate = true
		}

		if isCandidate {
			stats.CandidateReads++
			candidates = append(candidates, p.recordToStruct(rec))
		}
	}

	return candidates, stats, nil
}

// ExtractCandidateReadsParallel extracts candidates using multiple workers
func (p *BAMParser) ExtractCandidateReadsParallel(mapqThreshold int, includeUnmapped, includeLowMapQ bool) ([]*ReadRecord, *Stats, error) {
	f, err := os.Open(p.bamPath)
	if err != nil {
		return nil, nil, fmt.Errorf("opening BAM: %w", err)
	}
	defer f.Close()

	br, err := bam.NewReader(f, 0)
	if err != nil {
		return nil, nil, fmt.Errorf("creating BAM reader: %w", err)
	}
	defer br.Close()

	// Channels for parallel processing
	recordChan := make(chan *sam.Record, p.chunkSize*2)
	resultChan := make(chan *ReadRecord, p.chunkSize*2)
	var wg sync.WaitGroup

	// Start workers
	for i := 0; i < p.numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for rec := range recordChan {
				// Skip secondary and supplementary
				if rec.Flags&sam.Secondary != 0 || rec.Flags&sam.Supplementary != 0 {
					continue
				}

				// Check if candidate
				isCandidate := false
				if includeUnmapped && rec.Flags&sam.Unmapped != 0 {
					isCandidate = true
				} else if includeLowMapQ && rec.Flags&sam.Unmapped == 0 && int(rec.MapQ) < mapqThreshold {
					isCandidate = true
				}

				if isCandidate {
					resultChan <- p.recordToStruct(rec)
				}
			}
		}()
	}

	// Collector goroutine
	var candidates []*ReadRecord
	var collectWg sync.WaitGroup
	collectWg.Add(1)
	go func() {
		defer collectWg.Done()
		for rec := range resultChan {
			candidates = append(candidates, rec)
		}
	}()

	// Read records and send to workers
	stats := &Stats{}
	for {
		rec, err := br.Read()
		if err != nil {
			break
		}

		stats.TotalReads++

		if rec.Flags&sam.Unmapped != 0 {
			stats.UnmappedReads++
		} else {
			stats.MappedReads++
			if int(rec.MapQ) < mapqThreshold {
				stats.LowMapQReads++
			}
		}

		if p.extractBarcode(rec) != "" {
			stats.ReadsWithBarcode++
		}
		if p.extractUMI(rec) != "" {
			stats.ReadsWithUMI++
		}

		recordChan <- rec
	}

	close(recordChan)
	wg.Wait()
	close(resultChan)
	collectWg.Wait()

	stats.CandidateReads = int64(len(candidates))

	return candidates, stats, nil
}

// WriteToFastq writes records to FASTQ file
func WriteToFastq(records []*ReadRecord, outputPath string, compress bool) error {
	f, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer f.Close()

	var w *bufio.Writer
	if compress {
		gw := gzip.NewWriter(f)
		defer gw.Close()
		w = bufio.NewWriter(gw)
	} else {
		w = bufio.NewWriter(f)
	}
	defer w.Flush()

	for _, rec := range records {
		w.WriteString(rec.ToFastq())
	}

	return nil
}

// WriteToFasta writes records to FASTA file
func WriteToFasta(records []*ReadRecord, outputPath string, compress bool) error {
	f, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer f.Close()

	var w *bufio.Writer
	if compress {
		gw := gzip.NewWriter(f)
		defer gw.Close()
		w = bufio.NewWriter(gw)
	} else {
		w = bufio.NewWriter(f)
	}
	defer w.Flush()

	for _, rec := range records {
		w.WriteString(rec.ToFasta())
	}

	return nil
}

// WriteBarcodes writes unique barcodes to file
func WriteBarcodes(records []*ReadRecord, outputPath string) error {
	barcodes := make(map[string]bool)
	for _, rec := range records {
		if rec.Barcode != "" {
			barcodes[rec.Barcode] = true
		}
	}

	f, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer f.Close()

	w := csv.NewWriter(f)
	w.Comma = '\t'
	defer w.Flush()

	for barcode := range barcodes {
		w.Write([]string{barcode})
	}

	return nil
}

func main() {
	var (
		bamPath        = flag.String("i", "", "Input BAM file")
		outputFastq    = flag.String("o-fastq", "", "Output FASTQ file")
		outputFasta    = flag.String("o-fasta", "", "Output FASTA file")
		outputBarcodes = flag.String("o-barcodes", "", "Output barcodes file")
		barcodeTag     = flag.String("barcode-tag", "CB", "Cell barcode tag")
		umiTag         = flag.String("umi-tag", "UB", "UMI tag")
		mapqThreshold  = flag.Int("mapq", 30, "MAPQ threshold")
		chunkSize      = flag.Int("chunk-size", 10000, "Chunk size")
		numWorkers     = flag.Int("threads", 0, "Number of workers (0 = auto)")
		parallel       = flag.Bool("parallel", false, "Use parallel processing")
		compress       = flag.Bool("compress", false, "Compress output")
	)
	flag.Parse()

	if *bamPath == "" {
		log.Fatal("Please provide input BAM file with -i")
	}

	parser := NewBAMParser(*bamPath, *barcodeTag, *umiTag, *chunkSize, *numWorkers)

	log.Printf("Parsing BAM file: %s", *bamPath)

	var candidates []*ReadRecord
	var stats *Stats
	var err error

	if *parallel {
		candidates, stats, err = parser.ExtractCandidateReadsParallel(*mapqThreshold, true, true)
	} else {
		candidates, stats, err = parser.ExtractCandidateReads(*mapqThreshold, true, true)
	}

	if err != nil {
		log.Fatalf("Error parsing BAM: %v", err)
	}

	log.Printf("Parsing complete:")
	log.Printf("  Total reads: %d", stats.TotalReads)
	log.Printf("  Mapped reads: %d", stats.MappedReads)
	log.Printf("  Unmapped reads: %d", stats.UnmappedReads)
	log.Printf("  Low MAPQ reads: %d", stats.LowMapQReads)
	log.Printf("  Reads with barcode: %d", stats.ReadsWithBarcode)
	log.Printf("  Reads with UMI: %d", stats.ReadsWithUMI)
	log.Printf("  Candidate reads: %d", stats.CandidateReads)

	// Write outputs
	if *outputFastq != "" {
		log.Printf("Writing FASTQ: %s", *outputFastq)
		if err := WriteToFastq(candidates, *outputFastq, *compress); err != nil {
			log.Fatalf("Error writing FASTQ: %v", err)
		}
	}

	if *outputFasta != "" {
		log.Printf("Writing FASTA: %s", *outputFasta)
		if err := WriteToFasta(candidates, *outputFasta, *compress); err != nil {
			log.Fatalf("Error writing FASTA: %v", err)
		}
	}

	if *outputBarcodes != "" {
		log.Printf("Writing barcodes: %s", *outputBarcodes)
		if err := WriteBarcodes(candidates, *outputBarcodes); err != nil {
			log.Fatalf("Error writing barcodes: %v", err)
		}
	}

	log.Println("Done!")
}
